sim/midas/src/main/scala/midas/core/FPGATop.scala

```diff
@@ -406,41 +425,78 @@ class FPGATopImp(outer: FPGATop)(implicit p: Parameters) extends LazyModuleImp(o
-    "CTRL_BEAT_BYTES"-> ctrl.nastiWStrobeBits,
-    "CTRL_AXI4_SIZE" -> log2Ceil(ctrl.nastiWStrobeBits),
-    // These specify channel widths; used mostly in the test harnesses
-    "MEM_NUM_CHANNELS" -> p(HostMemNumChannels),
-    "MEM_ADDR_BITS"  -> p(HostMemChannelKey).axi4BundleParams.addrBits,
-    "MEM_DATA_BITS"  -> p(HostMemChannelKey).axi4BundleParams.dataBits,
-    "MEM_ID_BITS"    -> p(HostMemChannelKey).axi4BundleParams.idBits,
-    "MEM_STRB_BITS"  -> p(HostMemChannelKey).axi4BundleParams.dataBits / 8,
-    "MEM_BEAT_BYTES" -> p(HostMemChannelKey).axi4BundleParams.dataBits / 8,
-    // These are fixed by the AXI4 standard, only used in SW DRAM model
-    "MEM_SIZE_BITS"  -> AXI4Parameters.sizeBits,
-    "MEM_LEN_BITS"   -> AXI4Parameters.lenBits,
-    "MEM_RESP_BITS"  -> AXI4Parameters.respBits,
-    // Address width of the aggregated host-DRAM space
-    "CPU_MANAGED_AXI4_ID_BITS"    -> cpu_managed_axi4.map(_.params.idBits)      .getOrElse(0).toLong,
-    "CPU_MANAGED_AXI4_ADDR_BITS"  -> cpu_managed_axi4.map(_.params.addrBits)    .getOrElse(0).toLong,
-    "CPU_MANAGED_AXI4_DATA_BITS"  -> cpu_managed_axi4.map(_.params.dataBits)    .getOrElse(0).toLong,
-    "CPU_MANAGED_AXI4_STRB_BITS"  -> cpu_managed_axi4.map(_.params.dataBits / 8).getOrElse(0).toLong,
-    "CPU_MANAGED_AXI4_BEAT_BYTES" -> cpu_managed_axi4.map(_.params.dataBits / 8).getOrElse(0).toLong,
-    // Widths of the AXI4 FPGA to CPU channel
-    "FPGA_MANAGED_AXI4_ID_BITS"    -> fpga_managed_axi4.map(_.params.idBits)  .getOrElse(0).toLong,
-    "FPGA_MANAGED_AXI4_ADDR_BITS"  -> fpga_managed_axi4.map(_.params.addrBits).getOrElse(0).toLong,
-    "FPGA_MANAGED_AXI4_DATA_BITS"  -> fpga_managed_axi4.map(_.params.dataBits).getOrElse(0).toLong,
-    "FPGA_MANAGED_AXI4_STRB_BITS"  -> fpga_managed_axi4.map(_.params.dataBits / 8).getOrElse(0).toLong,
-    "FPGA_MANAGED_AXI4_BEAT_BYTES" -> fpga_managed_axi4.map(_.params.dataBits / 8).getOrElse(0).toLong,
-    // Data chunk width of LoadMemWidget
-    "MEM_DATA_CHUNK" -> outer.loadMem.memDataChunk
-  ) ++:
-   cpu_managed_axi4.map { _ => "CPU_MANAGED_AXI4_PRESENT" -> 1.toLong } ++:
-   fpga_managed_axi4.map { _ => "FPGA_MANAGED_AXI4_PRESENT" -> 1.toLong } ++:
-   Seq.tabulate[(String, Long)](p(HostMemNumChannels))(idx => s"MEM_HAS_CHANNEL${idx}" -> 1)
-  def genHeader(sb: StringBuilder)(implicit p: Parameters) = outer.genHeader(sb)
+  val confCtrl = AXI4Config(
+      ctrl.nastiXIdBits,
+      ctrl.nastiXAddrBits,
+      ctrl.nastiXDataBits,
+      ctrl.nastiWStrobeBits,
+      ctrl.nastiWStrobeBits
+  )
+  val confMem = AXI4Config(p(HostMemChannelKey).axi4BundleParams)
+  val confCPUManaged = cpu_managed_axi4.map(m => AXI4Config(m.params))
+  val confFPGAManaged = fpga_managed_axi4.map(m => AXI4Config(m.params))
+
+  def genHeader(sb: StringBuilder)(implicit p: Parameters) = {
+    outer.genHeader(sb)
+
+    sb append "#include \"emul/axi4.h\"\n"
+
+    def printConfig(conf: AXI4Config) {
+      sb append "AXI4Config{"
```

Comment:
style nit: Avoid `.`-less `infix` notation this has fallen out of vogue in 2.13+, 3.+